### PR TITLE
feat: change ArgumentList name to FuncParameterList

### PR DESCRIPTION
**Pull request by scgkiran**
Fixed bugs in matching variadic arguments.

BREAKING CHANGE: 
- `ArgumentList` has been renamed to `FuncParameterList`. 

BREAKING CHANGE: 
- The `ReturnType` method in the `FuncDefArgType` interface now requires `FuncParameterList` and actual argument types as parameters.

### CODE REVIEW

**Comment by scgkiran** [comment]
No logical change, just cleanup and renaming of variables.

File: extensions/simple_extension.go [hunk 1]
```diff
@@ -55,9 +56,9 @@ type TypeVariation struct {
 	Functions   TypeVariationFunctions
 }
 
-type Argument interface {
+type FuncParameter interface {
 	toTypeString() string
-	argumentMarker() // unexported marker method
```
Comment (jacques-n) [inline extensions/simple_extension.go:60]: why remove the marker? We should keep to encapsulate the set of valid items.

File: extensions/variants.go [hunk 2]
```diff
@@ -72,33 +72,38 @@ func validateType(arg Argument, actual types.Type, idx int, nullHandling Nullabi
...
+	funcParameters FuncParameterList, variadic *VariadicBehavior, argumentTypes []types.Type) (types.Type, error) {
+	if len(funcParameters) != len(argumentTypes) {
 		if variadic == nil {
 			return nil, fmt.Errorf("%w: mismatch in number of arguments provided. got %d, expected %d",
-				substraitgo.ErrInvalidExpr, len(actualTypes), len(paramTypeList))
+				substraitgo.ErrInvalidExpr, len(argumentTypes), len(funcParameters))
 		}
 
-		if !variadic.IsValidArgumentCount(len(actualTypes) - len(paramTypeList) - 1) {
+		if !variadic.IsValidArgumentCount(len(argumentTypes) - len(funcParameters) - 1) {
```
Comment (jacques-n) [inline extensions/variants.go:87]: Is this code right? It looks wrong.

File: types/types.go [hunk 3]
```diff
@@ -689,7 +781,7 @@ func (s *PrimitiveType[T]) MatchWithoutNullability(ot Type) bool {
 	return false
 }
 
-func (s *PrimitiveType[T]) ReturnType() (Type, error) {
+func (s *PrimitiveType[T]) ReturnType([]FuncDefArgType, []Type) (Type, error) {
```
Comment (jacques-n) [inline types/types.go:784]: not sure what this signature change is throughout the patch.

File: types/parameterized_list_type_test.go [hunk 4]
```diff
@@ -7,39 +7,39 @@ import (
 
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
-	"github.com/substrait-io/substrait-go/types"
+	. "github.com/substrait-io/substrait-go/types"
```
Comment (jacques-n) [inline types/parameterized_list_type_test.go:10]: why did you add this cosmetic changes in this patch in multiple places?

File: types/types.go [hunk 5]
```diff
@@ -431,6 +431,98 @@ type (
 	}
 )
 
+type EnumType struct {
```
Comment (jacques-n) [inline types/types.go:434]: docstring

File: types/types.go (hunk 3 above)
Comment (scgkiran) [reply types/types.go:791]: Raised PR #80 for return type derivation

**Review by EpsilonPrime — APPROVED** [review]
Some of the changed lines have no coverage but that's not related to this name change PR.

File: extensions/simple_extension.go (hunk 1 above)
Comment (jacques-n) [reply extensions/simple_extension.go:60]: it doesn't accomplish the same. The unexported method guarantees that others can't implement additional implementations outside the package.

Comment (scgkiran) [reply extensions/simple_extension.go:60]: Restored the marker method. 
Even without the marker method, this interface has one other unexported method(toTypeString).

File: extensions/variants.go (hunk 2 above)
Comment (scgkiran) [reply extensions/variants.go]: forgot to mention this is fixed added some testcases also.

**Comment by jacques-n** [comment]
This should have a breaking change commit message, right?

**Comment by scgkiran** [comment]
> This should have a breaking change commit message, right?

Updated the PR title and description.
